sdk/core/System.ClientModel/tests/client/ClientShared/ModelReaderWriterExtensions.cs

```diff
@@ -176,15 +176,22 @@ public static void WriteNumberValue(this Utf8JsonWriter writer, DateTimeOffset v
         writer.WriteNumberValue(value.ToUnixTimeSeconds());
     }
 
-    public static void WriteObjectValue(this Utf8JsonWriter writer, object? value)
+    private static IJsonModel<T> GetJsonInterface<T>(this ModelReaderWriterOptions options, IJsonModel<T> model)
+    {
+        options.ProxiedModel = model;
+        return options.Proxies.TryGetValue(model.GetType(), out var proxy) && proxy is IJsonModel<T> proxyAsT ? proxyAsT : model;
+    }
+
+    public static void WriteObjectValue(this Utf8JsonWriter writer, object? value, ModelReaderWriterOptions? options = null)
     {
         switch (value)
         {
             case null:
                 writer.WriteNullValue();
                 break;
             case IJsonModel<object> writeable:
-                writeable.Write(writer, ModelReaderWriterHelper.WireOptions);
```

Comment:
I wonder if we could change this line to ModelReaderWrite.Write(writable, options) instead of `options.GetJsonInterface(writeable).Write(writer, options)`?  Moving this direction would mean we keep a complex implementation pattern internal, and I believe we could make the GetJsonInterface and/or Proxies dictionary APIs internal.